Sources/Wendy/utils/FetchReleases.swift

```diff
@@ -261,39 +270,201 @@ func downloadAsset(_ asset: Release.Asset) async throws -> URL {
 enum ExtractError: Error {
     case failedToExtract
     case executableNotFound
+    case maliciousArchive(String)
 }
 
 func extract(
     at url: URL,
     to tempDir: URL,
     findExecutable: (URL) -> Bool
 ) async throws -> URL {
+    let logger = Logger(label: "sh.wendy.utils.fetchReleases")
+
     // Determine if it's a tar.gz or a binary
     let isTarGz = url.pathExtension == "gz" || url.lastPathComponent.contains("tar.gz")
     guard isTarGz else {
-        print("File is not a tar.gz file")
+        logger.debug(
+            "File is not a tar.gz file, returning as-is",
+            metadata: ["path": "\(url.path)"]
+        )
         return url
     }
 
-    print("Extracting tar.gz file...")
+    logger.info("Extracting tar.gz archive", metadata: ["path": "\(url.path)"])
     let extractDir = tempDir.appendingPathComponent("extract")
     try FileManager.default.createDirectory(at: extractDir, withIntermediateDirectories: true)
+
+    // Validate tar contents before extraction to prevent path traversal attacks
+    logger.debug("Validating tar archive contents")
+    let listProcess = Process()
+    listProcess.executableURL = URL(fileURLWithPath: "/usr/bin/env")
+    listProcess.arguments = ["tar", "-tzf", url.path]
+
+    let pipe = Pipe()
+    listProcess.standardOutput = pipe
+    let errorPipe = Pipe()
+    listProcess.standardError = errorPipe
+
+    try listProcess.run()
+    listProcess.waitUntilExit()
+
+    guard listProcess.terminationStatus == 0 else {
+        let errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()
+        let errorMessage = String(data: errorData, encoding: .utf8) ?? "unknown error"
+        logger.error(
+            "Failed to list tar archive contents",
+            metadata: [
+                "exit_code": "\(listProcess.terminationStatus)",
+                "archive": "\(url.path)",
+                "error": "\(errorMessage)",
+            ]
+        )
+        throw ExtractError.failedToExtract
+    }
+
+    let outputData = pipe.fileHandleForReading.readDataToEndOfFile()
+    let output = String(data: outputData, encoding: .utf8) ?? ""
+    let files = output.components(separatedBy: .newlines).filter { !$0.isEmpty }
+
+    // Check each file path for security issues
+    for filePath in files {
+        // Check for absolute paths
+        if filePath.hasPrefix("/") {
+            logger.error(
+                "Archive contains absolute path",
+                metadata: [
+                    "archive": "\(url.path)",
+                    "malicious_path": "\(filePath)",
+                ]
+            )
+            throw ExtractError.maliciousArchive("Archive contains absolute path: \(filePath)")
+        }
+
+        // Check for path traversal attempts
+        if filePath.contains("../") || filePath.contains("/..") {
```

Comment:
```suggestion
        if filePath.contains("..") {
```

Doesn't this make more sense? Users could still write "just" `..` or maybe work around it on Windows using `..\\` or some other trick